FurkannOzbek/application-hub
# [FEATURE] Implement GET/POST/DELETE for Notes

# Pull Request Title

**Example:** [FEATURE] Implement User Authentication

---

## Description

This PR implements new endpoints that allow to create/update, get, and delete notes associated with their applications. Each application can have one unique note, providing users with the ability to maintain additional context or comments for their applications.

Problem: Users needed a way to add and track notes for their applications
Impact: Users can now maintain contextual information for each application through a one-to-one note relationship

---

## Changes Made

### Backend:
- Implemented new controller `noteController.js` with CRUD operations
- Created DTO for consistent note data transformation
- Added new routes for note management under `user` router

### Database:
- Added `UNIQUE` constraint to `application_id` in `application_note` table

### API Endpoints:
- GET /applications/:id/note - Retrieve note for an application
- POST /applications/:id/note - Create/Update note (uses upsert)
- DELETE /applications/:id/note - Remove note

Error Handling:
- Implemented granular error handling with separate try-catch blocks for request processing and database operations

---

## Testing Steps

1. Create Note:
- Send POST request to `/applications/:id/note` with content
- Verify 201 response with note data
- Try creating another note for same application to verify upsert

2. Get Note:
- Send GET request to `/applications/:id/note`
- Verify correct note data is returned
- Test with non-existent application ID

3. Delete Note:
- Send DELETE request to `/applications/:id/note`
- Verify successful deletion
- Verify 404 when trying to delete non-existent note

4. Error Cases:
- Test with invalid application IDs
- Test with missing content
- Test with expired token

---

## Issues Addressed

Closes [Backend: Add endpoints for getting, adding and updating Notes](https://trello.com/c/MwOw9Jn8)
